### ISSUE TITLE
Slow shutdown and not cleanly closing listening sockets

### ISSUE THREAD

**Issue by nmaludy** [cutelyst/cutelyst#301]
I noticed two things when trying out Cutelyst:

#### 1. The server takes a _long_ time to shut down (~30 seconds).
Doing some investigation / debugging this only happens a user has initiated a connection to the server (example with the browser). In this case the `TcpServer` never closes its sockets, they stay open and after the timeout the `master` process has to to `SIGKILL` the worker children.

Doing some testing at the fix for this is (i think, feedback appreciated) is adding a call to `socket->connectionClose()` here:
https://github.com/cutelyst/cutelyst/blob/master/server/tcpserver.cpp#L87-L109

Example:
```cpp
void TcpServer::shutdown()
{
    if (isListening()) {
        pauseAccepting();
    }

    if (m_processing == 0) {
        m_engine->serverShutdown();
    } else {
        const auto childrenL = children();
        for (auto child : childrenL) {
            auto socket = qobject_cast<TcpSocket*>(child);
            if (socket) {
                socket->protoData->headerConnection = ProtocolData::HeaderConnectionClose;
                connect(socket, &TcpSocket::finished, this, [this] () {
                    if (!m_processing) {
                        m_engine->serverShutdown();
                    }
                });
               socket->connectionClose();
            }
        }
    }
}
```

#### 2. The server is not properly closing the file descriptor for the socket

Sometimes (only after a user has initiated a connection) and you shut down the server, then try to restart it, the TCP socket is unable to be opened with the error (notice i turned on native socket debugging `#define`):

```shell
2118788:2118788 wsgi.tcp_server_balancer[debug] QNativeSocketEnginePrivate::createNewSocket(1, 2) == true
2118788:2118788 wsgi.tcp_server_balancer[debug] QNativeSocketEnginePrivate::nativeBind(0.0.0.0, 3000) == false (Address already in use)
2118788:2118788 wsgi.tcp_server_balancer[critical] Failed to bind to socket 8
Failed to listen on TCP: :3000 : 
```

I have yet to track down the cause of the issue, but my gut says that the TcpServer is not cleanly closing the listening socket, causing in a bind failure when you launch the application again. Eventually the kernel does something under the hood, cleaning up the socket, and you can launch the server once more.

Some thoughts from reading/googleing, it might be necessary to:
- pause accepting connections (dont' accept new connections)
- read/write any outstanding buffers
- close the file desriptor

**Comment by dantti** [comment]
Yes, if there is an active connection and no timeout set it will hang for a while, the reason is that there might be some request running, closing the socket can be done by killing it.
The current code should pause new connections, but after that I think some handling logic is still needed, if you are on Linux I highly recommend using reuse-port, it is not only better for performance but also solves the failed to bind issue.

**Comment by dantti** [comment]
Fixes second issue 490d3fe1f3c4d24b384b38ca79605de51ac64a32
